src/securemul.rs

```diff
@@ -0,0 +1,430 @@
+    use rand_core::RngCore;
+
+    use crate::replicated_secret_sharing::ReplicatedSecretSharing;
+
+    use futures::{stream, StreamExt};
+    use futures_util::future::join_all;
+
+    use crate::prss::Participant;
+
+    use crate::error::BoxError;
+    use crate::helpers;
+    use crate::helpers::ring::mock::TestHelper;
+    use crate::securemul::stream::secure_multiply;
+    use crate::securemul::{ProtocolContext, SecureMul};
+
+    #[tokio::test]
+    async fn basic() -> Result<(), BoxError> {
+        let ring = helpers::ring::mock::make_three();
+        let participants = crate::prss::test::make_three();
+        let context = make_context(&ring, &participants);
+        let mut rand = StepRng::new(1, 1);
+
+        assert_eq!(30, multiply_sync(&context, 6, 5, &mut rand).await?);
+        assert_eq!(25, multiply_sync(&context, 5, 5, &mut rand).await?);
+        assert_eq!(7, multiply_sync(&context, 7, 1, &mut rand).await?);
+        assert_eq!(0, multiply_sync(&context, 0, 14, &mut rand).await?);
+        assert_eq!(8, multiply_sync(&context, 7, 10, &mut rand).await?);
+        assert_eq!(4, multiply_sync(&context, 5, 7, &mut rand).await?);
+        assert_eq!(1, multiply_sync(&context, 16, 2, &mut rand).await?);
+
+        Ok(())
+    }
+
+    /// Secure multiplication may be used with Stream API where shares are provided as elements
+    /// of a `Stream`.
+    #[tokio::test]
+    async fn supports_stream_of_secret_shares() {
+        // we compute a*b*c in this test. 4*3*2 = 24
+        let mut rand = StepRng::new(1, 1);
+        let a = share(Fp31::from(4_u128), &mut rand);
+        let b = share(Fp31::from(3_u128), &mut rand);
+        let c = share(Fp31::from(2_u128), &mut rand);
+        let start_index = 1024_u128;
+
+        // setup helpers
+        let ring = helpers::ring::mock::make_three();
+        let participants = crate::prss::test::make_three();
+        let participants = [participants.0, participants.1, participants.2];
+
+        // dedicated streams for each helper
+        let input = [
+            stream::iter(vec![a[0], b[0], c[0]]),
+            stream::iter(vec![a[1], b[1], c[1]]),
+            stream::iter(vec![a[2], b[2], c[2]]),
+        ];
+
+        // create 3 tasks (1 per helper) that will execute secure multiplication
+        let handles = input.into_iter().zip(participants).zip(ring).map(
+            |((input, participant), helper_ring)| {
+                tokio::spawn(async move {
+                    let ctx = ProtocolContext {
+                        participant: &participant,
+                        helper_ring: &helper_ring,
+                    };
+                    let mut stream = secure_multiply(input, &ctx, start_index);
+
+                    // compute a*b
+                    let _ = stream.next().await.expect("Failed to compute a*b");
+
+                    // compute (a*b)*c and return it
+                    stream.next().await.expect("Failed to compute a*b*c")
```

Comment:
Is the intention for protocol authors to write code like this? I don't find this very transparent. It's not clear that this computes a*b*c - could we make that shine through more?

Reply:
OK, I believe you.